collect_app/src/test/java/org/odk/collect/android/formmanagement/ServerFormDownloaderTest.java

```diff
@@ -0,0 +1,349 @@
+        assertThat(new String(read(mediaFile2)), is("contents2"));
+    }
+
+    @Test
+    public void afterDownloadingXForm_cancelling_throwsInterruptedExceptionAndDoesNotSaveAnything() throws Exception {
+        String xform = createXForm("id", "version");
+        ServerFormDetails serverFormDetails = new ServerFormDetails(
+                "Form",
+                "http://downloadUrl",
+                "http://manifestUrl",
+                "id",
+                "version",
+                "md5:" + FileUtils.getMd5Hash(new ByteArrayInputStream(xform.getBytes())),
+                true,
+                false,
+                null);
+
+        CancelAfterFormDownloadFormListApi formListApi = new CancelAfterFormDownloadFormListApi(xform);
+        ServerFormDownloader downloader = new ServerFormDownloader(formListApi, formsRepository, cacheDir, formsDir.getAbsolutePath());
+
+        try {
+            downloader.downloadForm(serverFormDetails, null, formListApi);
+            fail("Excepted exception");
+        } catch (InterruptedException e) {
+            assertThat(formsRepository.getAll(), is(empty()));
+            assertThat(asList(new File(getCacheFilesPath()).listFiles()), is(empty()));
+            assertThat(asList(new File(getFormFilesPath()).listFiles()), is(empty()));
+        }
+    }
+
+    @Test
+    public void afterDownloadingMediaFile_cancelling_throwsInterruptedExceptionAndDoesNotSaveAnything() throws Exception {
+        String xform = createXForm("id", "version");
+        ServerFormDetails serverFormDetails = new ServerFormDetails(
+                "Form",
+                "http://downloadUrl",
+                "http://manifestUrl",
+                "id",
+                "version",
+                "md5:" + FileUtils.getMd5Hash(new ByteArrayInputStream(xform.getBytes())),
+                true,
+                false,
+                new ManifestFile("", asList(
+                        new MediaFile("file1", "hash-1", "http://file1"),
+                        new MediaFile("file2", "hash-2", "http://file2")
+                )));
+
+        CancelAfterMediaFileDownloadFormListApi formListApi = new CancelAfterMediaFileDownloadFormListApi(xform);
+        ServerFormDownloader downloader = new ServerFormDownloader(formListApi, formsRepository, cacheDir, formsDir.getAbsolutePath());
+
+        try {
+            downloader.downloadForm(serverFormDetails, null, formListApi);
+            fail("Excepted exception");
+        } catch (InterruptedException e) {
+            assertThat(formsRepository.getAll(), is(empty()));
+            assertThat(asList(new File(getCacheFilesPath()).listFiles()), is(empty()));
+
+            // The media directory is created early for some reason
+            assertThat(asList(new File(getFormFilesPath()).listFiles()), contains(new File(getFormFilesPath() + "/Form-media")));
+        }
+    }
+
+    @Test
+    public void beforeDownloadingMediaFile_reportsProgress() throws Exception {
```

Comment:
Add test for media file progress?

Reply:
I'm pretty sure I ended up not needing another test to drive that out (I basically used commenting/deleting to work out adding tests). You're right that from a documentation and regression standpoint though it would be nice to have the test.

Reply:
Wait looking back at this I'm not sure what test you think is missing here? I thought I was missing media file progress but I think this test is just badly named as it covers that?